File: backend/services/kitchen-service/shared/auth/authorizer.js
/**
 * Shared Auth Authorizer para API Gateway
 * 
 * ⚠️ SEGURIDAD:
 * - NO incluye credenciales hardcodeadas
 * - Valida JWT con tenant_id obligatorio
 * - Compatible con AWS Academy
 */

const { verifyToken } = require('./jwt-utils');

/**
 * Lambda Authorizer para API Gateway
 * Valida JWT tokens en el header Authorization
 */
module.exports.handler = async (event) => {
  console.log('🔐 Auth Request:', {
    methodArn: event.methodArn,
    hasToken: !!event.authorizationToken,
    headers: event.headers
  });

  try {
    // Extraer token del header (REQUEST authorizer usa event.headers)
    let authHeader = event.authorizationToken;
    
    // Si no está en authorizationToken, buscar en headers
    if (!authHeader && event.headers) {
      authHeader = event.headers.Authorization || event.headers.authorization;
    }
    
    const token = extractToken(authHeader);
    
    if (!token) {
      console.error('❌ No token provided');
      throw new Error('No token provided');
    }

    // Verificar y decodificar JWT (incluye validación de tenant_id)
    const decoded = await verifyToken(token);
    
    console.log('✅ Token válido:', {
      userId: decoded.userId,
      role: decoded.role,
      tenant_id: decoded.tenant_id
    });

    // Validar tenant_id para usuarios staff (no Cliente)
    if (decoded.role !== 'Cliente' && !decoded.tenant_id) {
      console.error('❌ Staff user sin tenant_id');
      throw new Error('tenant_id requerido');
    }

    // Generar policy de acceso
    const policy = generatePolicy(decoded.userId, 'Allow', event.methodArn, {
      userId: decoded.userId,
      email: decoded.email,
      role: decoded.role,
      tenant_id: decoded.tenant_id || 'null'
    });
    
    console.log('📋 Policy generada:', JSON.stringify(policy, null, 2));
    
    return policy;

  } catch (error) {
    console.error('❌ Auth Error:', error.message);
    
    // Retornar Deny si el token es inválido
    throw new Error('Unauthorized');
  }
};

/**
 * Extraer token del header Authorization
 */
function extractToken(authHeader) {
  if (!authHeader) {
    return null;
  }

  // Formato: "Bearer <token>"
  const parts = authHeader.split(' ');
  
  if (parts.length === 2 && parts[0] === 'Bearer') {
    return parts[1];
  }

  // Si no tiene "Bearer", asumir que es solo el token
  return parts[0];
}

/**
 * Generar IAM Policy para API Gateway
 */
function generatePolicy(principalId, effect, resource, context) {
  const authResponse = {
    principalId
  };

  if (effect && resource) {
    // Usar wildcard para permitir acceso a todos los recursos del API
    // Extraer el ARN base del resource
    // De: arn:aws:execute-api:region:account:api-id/stage/METHOD/path/*
    // A:  arn:aws:execute-api:region:account:api-id/*
    const arnParts = resource.split(':');
    const apiGatewayArn = arnParts.slice(0, 5).join(':') + ':*';
    
    authResponse.policyDocument = {
      Version: '2012-10-17',
      Statement: [
        {
          Action: 'execute-api:Invoke',
          Effect: effect,
          Resource: apiGatewayArn
        }
      ]
    };
  }

  // Context se pasa a la función Lambda en event.requestContext.authorizer
  if (context) {
    authResponse.context = context;
  }

  return authResponse;
}

File: backend/shared/auth/authorizer.py
"""
Lambda Authorizer
Valida tokens JWT en requests a API Gateway
"""

import json
import sys
import os

# Agregar shared al path para imports
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..'))

from shared.auth.jwt_utils import decode_token, extract_token_from_header


def handler(event, context):
    """
    Lambda Authorizer para API Gateway
    Valida el token JWT y retorna una policy IAM
    """
    try:
        print(f"[Authorizer] Evento recibido: {json.dumps(event)}")
        
        # Extraer token según el tipo de authorizer
        # TOKEN authorizer: event['authorizationToken']
        # REQUEST authorizer: event['headers']['Authorization']
        if event.get('type') == 'TOKEN':
            authorization_header = event.get('authorizationToken', '')
        else:
            authorization_header = event.get('headers', {}).get('Authorization') or \
                                  event.get('headers', {}).get('authorization')
        
        token = extract_token_from_header(authorization_header)
        
        # Decodificar y validar token
        payload = decode_token(token)
        
        user_id = payload.get('userId')
        email = payload.get('email')
        role = payload.get('role')
        tenant_id = payload.get('tenantId')
        
        print(f"[Authorizer] Token válido para usuario {user_id} con rol {role}")
        
        # Generar policy IAM para permitir acceso a todos los endpoints
        # Convertir el methodArn específico a un wildcard
        # De: arn:aws:execute-api:region:account:api-id/stage/method/path
        # A:  arn:aws:execute-api:region:account:api-id/stage/*/*
        method_arn_parts = event['methodArn'].split('/')
        base_arn = '/'.join(method_arn_parts[:2])  # arn:aws:.../stage
        wildcard_arn = f"{base_arn}/*/*"
        
        policy = generate_policy(user_id, 'Allow', wildcard_arn, {
            'userId': user_id,
            'email': email,
            'role': role,
            'tenantId': tenant_id or ''
        })
        
        return policy
        
    except Exception as e:
        print(f"[Authorizer] Error de autorización: {str(e)}")
        # Retornar Deny policy con wildcard
        method_arn_parts = event['methodArn'].split('/')
        base_arn = '/'.join(method_arn_parts[:2])
        wildcard_arn = f"{base_arn}/*/*"
        return generate_policy('unauthorized', 'Deny', wildcard_arn)


def generate_policy(principal_id, effect, resource, context=None):
    """
    Genera una policy IAM para API Gateway
    """
    policy = {
        'principalId': principal_id,
        'policyDocument': {
            'Version': '2012-10-17',
            'Statement': [
                {
                    'Action': 'execute-api:Invoke',
                    'Effect': effect,
                    'Resource': resource
                }
            ]
        }
    }
    
    # Agregar context con información del usuario
    if context:
        policy['context'] = context
    
    return policy
